state/fork/traverse.go

```diff
@@ -0,0 +1,117 @@
+package fork
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// functor that will be called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) error
+
+// TraverseBackward traverses the given fork (specified by block ID `forkHead`)
+// in the order of decreasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
+func TraverseBackward(headers storage.Headers, forkHead flow.Identifier, visitor onVisitBlock, terminal Terminal) error {
+	lowestHeightToVisit, sanityChecker, err := terminal.Translate2Height(headers)
+	if err != nil {
+		return fmt.Errorf("error evaluating terminal: %w", err)
+	}
+
+	startBlock, err := headers.ByBlockID(forkHead)
+	if err != nil {
+		return fmt.Errorf("could not fork head with if  %x: %w", forkHead, err)
+	}
+	if startBlock.Height < lowestHeightToVisit {
+		return nil
+	}
+
+	return unsafeTraverse(headers, startBlock, visitor, lowestHeightToVisit, sanityChecker)
+}
+
+// unsafeTraverse implements the fork traversal in the order of decreasing height.
+// It is unsafe because:
+// * always calls the `visitor` on the `block`, _before_ it checks whether to stop
+// * if `block` has a lower height than lowestHeightToVisit, the traversal keeps
+//   going until we hit the root block
```

Comment:
In fact, we could consider moving the pre-check into the `unsafeTraverse`, and remove the `unsafe` in the name. This could work, because unsafeTraverse takes `startBlock` and `lowestheightToVisit` anyway